src/compiler/checker.ts

```diff
@@ -31166,7 +31166,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
 
     function getContextualTypeForStaticPropertyDeclaration(declaration: PropertyDeclaration, contextFlags: ContextFlags | undefined): Type | undefined {
         const parentType = isExpression(declaration.parent) && getContextualType(declaration.parent, contextFlags);
-        if (!parentType) return undefined;
+        if (!parentType || getSymbolOfDeclaration(declaration.parent) === parentType.symbol) return undefined;

```

Comment:
The idea behind the fix is very similar to the one [here](https://github.com/microsoft/TypeScript/pull/58392). We can't get any new information from a situation like this so we can avoid resolving members of this class at this point.